### ISSUE TITLE
Run query with parameters

### ISSUE THREAD

**Issue by doekman** [jakob/Postico#297]
In my Python-code (with sqlalchemy and psycopg2), I have queries like:

    INSERT INTO my_table(title, isbn, copy_nr)
    VALUES(:title, :isbn, 1)

the parts `:title` and `:isbn` are parameters which are supplied by code. 

I would be very handy, when parameters are detected, Postico would pop-up a window to enter values for those parameters (types would be needed to be specified too). Ideally, Postico would recognise if the same query was run, so it could remember the entered values and provide them as default values.

I've used this behaviour in TOAD, and it makes a developers life a little better :-)

Not sure if this fits in your philosophy about Postico.

**Comment by jakob** [comment]
Yes, this sounds like a good idea, and a number of people have already requested this sort of feature.

Is a type really needed? Wouldn't PostgreSQL infer the type? And if you need a specific type, wouldn't you just do something like:

```sql
INSERT INTO my_table(title, isbn, copy_nr)
VALUES(:title::TEXT, :isbn::MY_CUSTOM_TYPE, 1)
```

**Comment by doekman** [comment]
Python uses types for the queries; most types I use are text, integer, date/time (though at the moment date/time are ISO date strings; I'm new to Python). 

You could just ask for Postgres literals, so a difference between an empty string and NULL can be made (Microsoft came up with the solution to type Ctrl+0 to insert a NULL value in the UI, but I don't like it). And you don't need type-UI/type-dropdown's with this solution. To not bother the user, quotes can be added automatically (when it's not NULL/integer/already a string). 

I always think it's great to list some examples (input between brackets):

* []: empty string
* [NULL]: SQL's NULL
* [123]: All digits, it's probably an integer, nothing to do
* [12.3]: A fixed point number, also nothing to doo
* [+1.23e-4]: A float/real, nothing to do
* [Hello World's!]: A string, automatically escape and enclose in quotes
* ['Hi there']: A string, already quoted so nothing to infer
* [2016-11-28]: A string, because it's not all digits, so add quotes
* [TRUE]: A boolean, nothing to do
* [yes]: A string, so add quotes

Not sure if I want to add cast's (::TEXT) to my code, just for this feature.

Hope this helps :-)

**Comment by jakob** [comment]
Hm. I guess I could do it similar to how you input DEFAULT values in the structure editor: A text field and a popup. The popup lets you select if you want to input an expression or a literal value or NULL, and the text field would be interpreted accordingly (or hidden when you select NULL)

I really want Postico to handle the quoting, since it's such a pain to get this right. Would also make it a lot easier to enter arbitrary text into a query, without wondering if it is quoted or not.

**Comment by doekman** [comment]
I didn't think of that. Sounds that would work great!

**Comment by oakstreetsoftware** [comment]
Can't you just use the PQexecParams() to pass parameters, rather than worrying about which ones to quote or not quote?

**Comment by jakob** [comment]
@oakstreetsoftware: PQexecParams only supports a single query, doesn't tell me how many params are in a query, and I think it doesn't support params like :param (params need to be provided like $1, $2...)

**Comment by oakstreetsoftware** [comment]
Yes, you would need to parse and replace the query placeholders, but I assumed you'd need to do that, anyway before you could construct a dialog box with input fields.

I'm no sure I understand why supporting a single query is a problem. If you are parsing the query, yourself, so you can replace parameters, you could easily separate commands, if you support that.

Right now I'm just evaluating your software for myself and one of my clients. I have more questions about it than I have answers. I was only addressing the issue with quoting and escaping field values that were provided through a dynamic interface. It's not that different from a web-app, and I have a lot of experience with those.

Sorry, if my suggestions are not helpful.

**Comment by jakob** [comment]
@oakstreetsoftware Sorry if my response came across a bit harsh, I just wanted to summarise why I'm not using PQexecParams (I wanted to use it initially, but decided against it).

I do appreciate your suggestions, it's very valuable for me to get so much input!

**Comment by hraban** [comment]
@jakob on the other hand I'd like to add a vote for using PQexecParams precisely because it is a native feature which really gets sent on the wire, not just a client side enhancement.

In my case, I have some statements using params (i.e. with $1..) in my app, and I wanted to check their behaviour using Postico. I actually just tried what would happen if I just put a $1 in the query:

```sql
SELECT $1;
``` 

Perhaps it would immediately show an extra input field, allowing me to set the value of that param? :)

One of the things I use Postico for, is to sanity check the raw behaviour of PG. Sort of an "enhanced psql". This helps me check whether a problem is in my app logic, or with the way I use PG. Presently I can't do that for parameterised queries.

Of course, others might do things considerably differently. Just thought I'd share my 2 cents on the matter :)

**Comment by doekman** [comment]
Another reason for this feature, which is also using the `:name` format, [PugSQL](https://pugsql.org) which is a library for Python making it easy to use plain SQL files with parameters.

**Comment by sjmueller** [comment]
I also vote with @hraban with the native `$1` format, we are using Elixir and the queries that are generated use the `$1` `$2` format, I'm always dumping into the Postico SQL Query window and then manually replacing the params with actual values, it's a real pain when there are so many on a large query. Wish we could somehow run the query as is!

**Comment by ShiftyMcCool** [comment]
Any update on this feature?  I have no opinion on format as long as the functionality exists 😁

**Comment by yunyu** [comment]
Had to switch to TablePlus due to the lack of this feature :(

**Comment by airhorns** [comment]
Just me, another long time postico user, chiming in to say some solution to this would be really really great, it is super annoying to manually replace values in complicated parameterized queries!

**Comment by hraban** [comment]
> In my case, I have some statements using params (i.e. with $1..) in my app, and I wanted to check their behaviour using Postico. I actually just tried what would happen if I just put a $1 in the query:
> 
> ```sql
> SELECT $1;
> ```
> 
> Perhaps it would immediately show an extra input field, allowing me to set the value of that param? :)

It turns out DBeaver does exactly this

**Comment by doekman** [comment]
I've found a list with different parameter placeholder styles:

> String constant stating the type of parameter marker formatting expected by the interface. Possible values are*:
> 
> | paramstyle | Meaning                                                     |
> | ---------- | ----------------------------------------------------------- |
> | `qmark`      | Question mark style, e.g. `...WHERE name=?`                   |
> | `numeric`    | Numeric, positional style, e.g. `...WHERE name=:1`            |
> | `named`      | Named style, e.g. `...WHERE name=:name`                       |
> | `format`     | ANSI C printf format codes, e.g. `...WHERE name=%s`           |
> | `pyformat`   | Python extended format codes, e.g. `...WHERE name=%(name)s`   |
> 
> 
> *) Module implementors should prefer numeric, named or pyformat over the other formats because these offer more clarity and flexibility.

Found here: https://peps.python.org/pep-0249/#paramstyle

Not complete (it's obviously missing the `$1`-style), but hope this helps when you consider this feature 🙂

**Comment by bjartek** [comment]
I second the need for this. personally i used pgx/sqlc in golang and the namedArg syntax there is "select * from foo where name=@param_name"

**Comment by jakob** [comment]
I've tried to add support for query parameters to Postico! You can download development build 9788 or later to try the feature.

For now, I've only added support for these formats:

- `$1`, `$2` (standard PostgreSQL placeholders)
- `:name`

I have not added support for `?` placeholders, since that is the [JSON "key exists" operator](https://www.postgresql.org/docs/current/functions-json.html). If people really need ? style placeholders, I would need to think about ways that I could detect that it is a placeholder vs. an operator.

I have also not added support for `@param` placeholders, since these could interfere with operators that end in `@` (like the `<@` or `@@` JSON operators)

**Comment by jakob** [comment]
Development builds can be downloaded here: https://releases.eggerapps.at/postico2/changelog?update_channel=1

**Comment by doekman** [comment]
Nice 💯. The `:name` format works nicely with Python's SQLAlchemy. 

Some feedback from my initial testing.  This was actually the first example I ran (things happen to break when I get my hands on them 🐒):

```
SELECT * 
FROM generate_series(:from, :until)
;
```

When I supply `1` for `from` and `10` for `until` I get an error message:

```
ERROR:  function generate_series(unknown, unknown) is not unique
LINE 1: SELECT * from generate_series('1', '10');
                      ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
```

Apparently the parameters are always interpreted as `text`(you mentioned that in #339), so this is to be expected, for the moment at least.

For my use case, I can be using different types. The parameters in _SQLAlchemy_ are supplied by a Python `dict` structure (key/value's), and (SQL) types can be amongst `NULL`, `text`, `int`, `numeric`, `json` and maybe `boolean`. It can also be tuples (`select * from users where username in :usernames`) but that is really advanced.

So I hope every parameter input will get a type qualifier, and also something to distinguish between `NULL` and an empty string.

---

Another thing that breaks are type-casts:

```select '12'::int;```

Postico thinks `int` is a parameter. For you convenience, I [found the regular expression](https://github.com/sqlalchemy/sqlalchemy/blob/a96c607cc53f16b364d5025c5a5de1470661d3c9/lib/sqlalchemy/sql/elements.py#L2571) SQLAlchemy uses to collect the parameters for the query. Basically, this means (not sure you know your way in Python):

* `(?<![:\w\x5c])`: negative lookbehind assertion for the characters `[:\w\x5c]`
* `:(\w+)`: this is the name of the parameter itself, including the colon
* `(?!:)`: negative lookahead assertion for a colon.

Hope this helps.

---

Finally: I like it, input-values are persisted within the session, using the parameter-names as a unique key. I think this will do just fine.
